File: tools/tracer/src/tracer/commands/glob.py
"""`trace glob` — file-path pattern search.

Mirrors Claude Code's Glob tool: a pattern matched against full paths under a
base directory, with `**` for recursive descent. Returns the complete matched
file list, deterministically sorted by path, suitable as a strict replacement
for native Glob.

Default output is paths only. `--details` adds per-line ccn + rank + lifecycle
shoulder — the shape the deleted `trace context --glob` mode used to emit.

Gitignore is respected: the universe of candidate files comes from
`git ls-files --cached --others --exclude-standard`; pattern matching runs
against that universe. Outside a git repo, a SKIP_DIRS-bounded walk is the
fallback.
"""

from __future__ import annotations

import json
import os
import subprocess
from pathlib import Path

import click

from tracer import cache, file_facts, passive_context
from tracer.deps import require_dependencies


# Directories the fallback walker (non-git base) refuses to descend into.
# Matches the SKIP_DIRS used elsewhere in tracer for consistency.
_SKIP_DIRS = {
    ".git", "node_modules", ".venv", "venv", "__pycache__", "dist", "build",
    ".next", ".tracer-cache", ".pytest_cache", ".mypy_cache", ".ruff_cache",
    "vendor", "worktrees",
}


def _tracked_universe(base: Path) -> set[Path] | None:
    """Set of absolute paths git considers tracked-or-not-ignored under `base`.

    Returns None when `base` is not inside a git repo — caller falls back to
    the SKIP_DIRS walker.
    """
    try:
        result = subprocess.run(
            ["git", "ls-files", "--cached", "--others", "--exclude-standard"],
            cwd=base,
            capture_output=True,
            text=True,
            check=True,
            timeout=30,
        )
    except (subprocess.CalledProcessError, FileNotFoundError, subprocess.TimeoutExpired):
        return None
    return {(base / line).resolve() for line in result.stdout.splitlines() if line}


def _walk_universe(base: Path) -> set[Path]:
    """Fallback file universe when `base` is not in a git repo."""
    out: set[Path] = set()
    for current_root, dirs, files in os.walk(base):
        dirs[:] = [d for d in dirs if d not in _SKIP_DIRS and not d.startswith(".")]
        for name in files:
            out.add((Path(current_root) / name).resolve())
    return out


def _resolve_glob(pattern: str, base: Path) -> list[Path]:
    """Resolve `pattern` against `base` and intersect with the tracked universe.

    Returns absolute paths to regular files, deterministically sorted.
    """
    try:
        raw_matches = list(base.glob(pattern))
    except (OSError, ValueError):
        return []

    universe = _tracked_universe(base)
    ignore_policy = "gitignore"
    if universe is None:
        universe = _walk_universe(base)
        ignore_policy = "skip_dirs"

    matches = [
        path.resolve() for path in raw_matches
        if path.is_file() and path.resolve() in universe
    ]
    matches.sort()
    # Stash the policy on the function call site via attribute — caller
    # reads it back to populate the JSON envelope.
    _resolve_glob.last_ignore_policy = ignore_policy  # type: ignore[attr-defined]
    return matches


def _render_detail_line(path: Path, base: Path, repo_root: Path) -> dict:
    """Build the per-match detail payload (also reused in --json --details)."""
    try:
        rel = str(path.relative_to(base))
    except ValueError:
        rel = str(path)
    facts = file_facts.get(path, repo_root=repo_root)
    if facts is None:
        return {
            "path": rel,
            "ccn_total": 0,
            "rank": "unknown",
            "shoulder": None,
        }
    return {
        "path": rel,
        "ccn_total": facts.cyclomatic_complexity_total,
        "rank": facts.rank,
        "shoulder": passive_context.render_compact(facts),
    }


@click.command()
@click.argument("pattern")
@click.argument("base", required=False, default=".")
@click.option(
    "--details",
    is_flag=True,
    help="Per-line ccn + rank + lifecycle shoulder (vs bare paths by default).",
)
@click.option("--json", "as_json", is_flag=True, help="Machine-parseable JSON output.")
def command(pattern: str, base: str, details: bool, as_json: bool) -> None:
    """File-path pattern search with Claude Glob parity.

    Returns the complete file list matching `pattern` under `base`,
    deterministically sorted by path, respecting .gitignore inside a git repo.
    """
    require_dependencies()
    base_path = Path(base).resolve()
    if not base_path.exists():
        click.echo(f"Error: {base} does not exist", err=True)
        raise SystemExit(2)
    if not base_path.is_dir():
        click.echo(f"Error: {base} is not a directory", err=True)
        raise SystemExit(2)

    matches = _resolve_glob(pattern, base_path)
    ignore_policy = getattr(_resolve_glob, "last_ignore_policy", "gitignore")

    if as_json:
        repo_root = cache.repo_root_for(base_path)
        if details:
            entries = [_render_detail_line(p, base_path, repo_root) for p in matches]
            payload = {
                "pattern": pattern,
                "base": str(base_path),
                "ignore_policy": ignore_policy,
                "match_count": len(entries),
                "matches": entries,
            }
        else:
            rel_paths = []
            for p in matches:
                try:
                    rel_paths.append(str(p.relative_to(base_path)))
                except ValueError:
                    rel_paths.append(str(p))
            payload = {
                "pattern": pattern,
                "base": str(base_path),
                "ignore_policy": ignore_policy,
                "match_count": len(rel_paths),
                "matches": rel_paths,
            }
        click.echo(json.dumps(payload, indent=2))
        return

    if not matches:
        click.echo("(no matches)")
        return

    repo_root = cache.repo_root_for(base_path)
    if details:
        for path in matches:
            entry = _render_detail_line(path, base_path, repo_root)
            shoulder = entry["shoulder"] or ""
            click.echo(
                f"{entry['path']}  [ccn={entry['ccn_total']} {entry['rank']}]  {shoulder}"
            )
    else:
        for path in matches:
            try:
                click.echo(str(path.relative_to(base_path)))
            except ValueError:
                click.echo(str(path))

File: packages/claude/bin/tracer-dist/crate/src/commands/glob.rs
//! `trace glob` — Claude-Glob-shaped full-path pattern search.
//! `**` recurses, gitignore-respecting (universe = git ls-files),
//! deterministically sorted. Bare paths by default; `--details` adds
//! ccn + rank + lifecycle shoulder.

use crate::{cache, file_facts, passive_context};
use anyhow::Result;
use serde_json::{json, Value};
use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

const SKIP_DIRS: &[&str] = &[
    ".git",
    "node_modules",
    ".venv",
    "venv",
    "__pycache__",
    "dist",
    "build",
    ".next",
    ".tracer-cache",
    ".pytest_cache",
    ".mypy_cache",
    ".ruff_cache",
    "vendor",
    "worktrees",
];

/// Absolute resolved paths git considers tracked-or-not-ignored under
/// `base`, via the shared `git ls-files` enumeration (deleted-in-index
/// paths already excluded). None when not in a git repo.
fn tracked_universe(repo_root: &Path, base: &Path) -> Option<BTreeSet<PathBuf>> {
    crate::repo_files::tracked_paths(repo_root, Some(base)).map(|paths| {
        paths
            .into_iter()
            .map(|p| p.canonicalize().unwrap_or(p))
            .collect()
    })
}

/// SKIP_DIRS-bounded filesystem walk under `base` — the non-git fallback
/// universe.
fn walk_universe(base: &Path) -> BTreeSet<PathBuf> {
    let skip: BTreeSet<&str> = SKIP_DIRS.iter().copied().collect();
    let mut out = BTreeSet::new();
    let walker = walkdir::WalkDir::new(base).into_iter().filter_entry(|e| {
        if e.file_type().is_dir() {
            let name = e.file_name().to_string_lossy();
            e.path() == base
                || (!skip.contains(name.as_ref()) && !name.starts_with('.'))
        } else {
            true
        }
    });
    for entry in walker.flatten() {
        if entry.file_type().is_file() {
            let p = entry.path();
            out.insert(p.canonicalize().unwrap_or_else(|_| p.to_path_buf()));
        }
    }
    out
}

/// Resolve a glob against the universe. Returns (sorted absolute file
/// matches, ignore_policy).
///
/// Performance: the result is the glob match ∩ `universe`. `universe` is
/// already the gitignore/skip-dirs-pruned file set (no
/// `node_modules`/`vendor` walk), so rather than walking the filesystem
/// unboundedly, the pattern is compiled once with `globset` and tested
/// against each universe path's base-relative form. `globset` is the
/// pure-Rust matcher ripgrep uses — preserves the single-static-binary
/// property. Glob semantics: `/`-segmented, `**` = zero-or-more
/// directories, `*`/`?`/`[...]` never cross `/` (`literal_separator(true)`).
fn resolve_glob(pattern: &str, repo_root: &Path, base: &Path) -> (Vec<PathBuf>, &'static str) {
    let (universe, policy) = match tracked_universe(repo_root, base) {
        Some(u) => (u, "gitignore"),
        None => (walk_universe(base), "skip_dirs"),
    };
    let matcher = match build_glob(pattern) {
        Some(m) => m,
        // Unparseable pattern → no matches.
        None => return (Vec::new(), policy),
    };
    let base_resolved = base.canonicalize().unwrap_or_else(|_| base.to_path_buf());
    let mut set: BTreeSet<PathBuf> = BTreeSet::new();
    for p in &universe {
        if !p.is_file() {
            continue;
        }
        let rel = match p.strip_prefix(&base_resolved) {
            Ok(r) => r,
            Err(_) => continue,
        };
        if matcher.is_match(rel) {
            set.insert(p.clone());
        }
    }
    // BTreeSet already yields sorted, deduped order over the intersected set.
    (set.into_iter().collect(), policy)
}

/// Translate a glob pattern to a `globset::GlobMatcher`. `**` matches
/// zero-or-more path segments (globset `**` with `literal_separator(true)`
/// has that semantic), and a trailing `/**` matches every descendant. A
/// bare `**` matches everything. `*`/`?`/`[...]` are per-segment (never
/// cross `/`); `literal_separator(true)` enforces that.
fn build_glob(pattern: &str) -> Option<globset::GlobMatcher> {
    use globset::GlobBuilder;
    // Normalize: empty segments (leading/trailing/`//`) are ignored.
    let segs: Vec<&str> = pattern.split('/').filter(|s| !s.is_empty()).collect();
    if segs.is_empty() {
        return None;
    }
    let normalized = segs.join("/");
    GlobBuilder::new(&normalized)
        .literal_separator(true)
        .backslash_escape(true)
        .build()
        .ok()
        .map(|g| g.compile_matcher())
}

fn rel_to_base(path: &Path, base: &Path) -> String {
    path.strip_prefix(base)
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_else(|_| path.to_string_lossy().to_string())
}

pub fn run(pattern: &str, base: &str, details: bool, as_json: bool) -> Result<Value> {
    let abs = cache::absolutize(Path::new(base));
    if !abs.exists() {
        eprintln!("Error: {base} does not exist");
        std::process::exit(2);
    }
    if !abs.is_dir() {
        eprintln!("Error: {base} is not a directory");
        std::process::exit(2);
    }
    // Canonicalized base path, printed verbatim in output.
    let base_abs = abs.canonicalize().unwrap_or(abs);
    let base_path = base_abs.clone();

    let repo_root = cache::worktree_root_for(&base_abs).unwrap_or_else(|| cache::display_root(&base_abs));
    let (matches, ignore_policy) = resolve_glob(pattern, &repo_root, &base_abs);

    // Build the value once; the human view reads from it (no second
    // file_facts pass for `--details`).
    let value = if details {
        let entries: Vec<Value> = matches
            .iter()
            .map(|p| {
                let rel = rel_to_base(p, &base_abs);
                match file_facts::get(p, &repo_root, None) {
                    None => json!({
                        "path": rel,
                        "ccn_total": 0,
                        "rank": "unknown",
                        "shoulder": Value::Null,
                    }),
                    Some(f) => json!({
                        "path": rel,
                        "ccn_total": f.cyclomatic_complexity_total,
                        "rank": f.rank,
                        "shoulder": passive_context::render_compact(&f),
                    }),
                }
            })
            .collect();
        json!({
            "pattern": pattern,
            "base": base_path.to_string_lossy(),
            "ignore_policy": ignore_policy,
            "match_count": entries.len(),
            "matches": entries,
        })
    } else {
        let rels: Vec<String> =
            matches.iter().map(|p| rel_to_base(p, &base_abs)).collect();
        json!({
            "pattern": pattern,
            "base": base_path.to_string_lossy(),
            "ignore_policy": ignore_policy,
            "match_count": rels.len(),
            "matches": rels,
        })
    };

    // An empty result over a base that contains nested checkouts is a scope
    // fact, not an absence fact — name them so the next call is scoped inside.
    let mut value = value;
    let nested = if matches.is_empty() {
        crate::repo_files::nested_repo_rels(&base_abs)
    } else {
        Vec::new()
    };
    if !nested.is_empty() {
        value["nested_repos"] = json!(nested);
    }

    if as_json {
        return Ok(value);
    }

    if matches.is_empty() {
        println!("(no matches)");
        for r in &nested {
            println!("nested repository (its own search scope): {r}");
        }
        return Ok(value);
    }

    if details {
        for m in value["matches"].as_array().unwrap_or(&vec![]) {
            println!(
                "{}  [ccn={} {}]  {}",
                m["path"].as_str().unwrap_or(""),
                m["ccn_total"].as_i64().unwrap_or(0),
                m["rank"].as_str().unwrap_or(""),
                m["shoulder"].as_str().unwrap_or(""),
            );
        }
    } else {
        for m in value["matches"].as_array().unwrap_or(&vec![]) {
            println!("{}", m.as_str().unwrap_or(""));
        }
    }
    Ok(value)
}
